pkg/storage/bench_test.go

```diff
@@ -1588,35 +1597,96 @@ func runMVCCGarbageCollect(
 	// the returned slice that affects the oldest 'deleteVersions' versions. The
 	// first write for each key will be at `ts+(1,0)`, the second one
 	// at `ts+(1,1)`, etc.
-	// If numRangeKeys is set to 1 then range tombstone will be written at ts.
+	//
+	// Write 'numRangeKeys' covering random intervals of keys using set of keys
+	// from the set described above. One range key is always below all point keys.
+	// Range tombstones are spread evenly covering versions space inserted between
+	// every m point versions.
+	// Range tombstones use timestamps starting from 'ts+(0,1)` and increase wall
+	// clock between version and if there are more range tombstones than
+	// numVersions, then logical clock is used to distinguish between them.
 	//
 	// NB: a real invocation of MVCCGarbageCollect typically has most of the keys
 	// in sorted order. Here they will be ordered randomly.
+
+	pointKeyTs := func(index int) hlc.Timestamp {
+		return ts.Add(int64(index+1), 0)
+	}
+
+	rangeKeyTs := func(index, numVersions, numRangeKeys int) hlc.Timestamp {
+		wallTime := index * numVersions / numRangeKeys
+		base := wallTime * numRangeKeys / numVersions
+		logical := 1 + index - base
+		return ts.Add(int64(wallTime), int32(logical))
+	}
+
 	setup := func() (gcKeys []roachpb.GCRequest_GCKey) {
 		batch := eng.NewBatch()
-		if opts.numRangeKeys > 1 {
-			b.Fatal("Invalid bench data config. Number of range keys can be 0 or 1")
+		pointKeys := make([]roachpb.Key, opts.numKeys)
+		for i := 0; i < opts.numKeys; i++ {
+			pointKeys[i] = randutil.RandBytes(rng, opts.keyBytes)
+			gcKeys = append(gcKeys, roachpb.GCRequest_GCKey{
+				Timestamp: pointKeyTs(opts.deleteVersions - 1),
+				Key:       pointKeys[i],
+			})
 		}
-		if opts.numRangeKeys == 1 {
-			if err := MVCCDeleteRangeUsingTombstone(ctx, batch, nil, keys.LocalMax, keys.MaxKey,
-				ts, hlc.ClockTimestamp{}, nil, nil, true, 0, nil); err != nil {
+		// When keys are deleted using range tombstones MVCC operations perform
+		// smart checks to drop them altogether if there's no underlying data.
+		// We want to prevent that for testing purposes and we add a key at lower
+		// timestamp prior to putting first tombstone and subsequently remove it
+		// from the batch at the end of setup.
+		if opts.numRangeKeys > 0 {
+			if err := MVCCPut(ctx, batch, nil, keys.LocalMax, ts, hlc.ClockTimestamp{}, val, nil); err != nil {
```

Comment:
Why not simply set the `idempotent` parameter for `MVCCDeleteRangeUsingTombstone` to `false` instead? I don't think we want the idempotent behavior here at all, because we want to control the exact number of (defragmented) range keys written.

Reply:
TIL what idempotent for range tombstones is.

Reply:
Yeah, "idempotent" is really a bit of a misnomer here.